testing/architecture-test/src/test/java/org/gradle/architecture/test/PlatformBoundariesTest.java

```diff
@@ -0,0 +1,157 @@
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+import static org.gradle.architecture.test.ArchUnitFixture.freeze;
+import static org.gradle.architecture.test.ArchUnitFixture.safeReflect;
+
+@SuppressWarnings("unchecked")
+@AnalyzeClasses(packages = "org.gradle")
+public class PlatformBoundariesTest {
+
+    private static final List<Path> allPlatformDirs;
+    private static final Map<String, List<Path>> dirsByPlatform;
+    private static final Map<String, List<Path>> usedDirsByPlatform;
+    private static final Map<String, List<Path>> allowedDirsByPlatform;
+
+    static {
+        Path basePath = Paths.get(System.getProperty("org.gradle.architecture.platforms-base-path"));
+        Path jsonPath = Paths.get(System.getProperty("org.gradle.architecture.platforms-json"));
+        Map<String, Map<String, List<String>>> data;
+        try (Reader reader = new FileReader(jsonPath.toFile())) {
+            data = (Map<String, Map<String, List<String>>>) new Gson().fromJson(reader, Map.class);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        allPlatformDirs = data.values().stream().flatMap(
+            info -> resolvePlatformDirs(basePath, info.get("dirs"))
+        ).collect(toList());
+        dirsByPlatform = data.entrySet().stream().collect(toMap(
+            Map.Entry::getKey,
+            v -> resolvePlatformDirs(basePath, v.getValue().get("dirs")).collect(toList())
+        ));
+        usedDirsByPlatform = data.entrySet().stream().collect(toMap(
+            Map.Entry::getKey,
+            v -> v.getValue().get("uses").stream().flatMap(
+                use -> dirsByPlatform.get(use).stream()
+            ).distinct().collect(toList())
+        ));
+        allowedDirsByPlatform = data.entrySet().stream().collect(toMap(
+            Map.Entry::getKey,
+            v -> Stream.concat(
+                usedDirsByPlatform.getOrDefault(v.getKey(), allPlatformDirs).stream(),
+                dirsByPlatform.get(v.getKey()).stream()
+            ).distinct().collect(toList())
+        ));
+    }
+
+    @ArchTest
+    public static final ArchRule platform_coupling = freeze(classes()
+        .that(areInPlatformDirectories())
+        .should(importClassesOnlyFromAllowedPlatforms()));
+
+    private static DescribedPredicate<? super JavaClass> areInPlatformDirectories() {
+        return new DescribedPredicate<JavaClass>("are in platform directories") {
+            @Override
+            public boolean test(JavaClass javaClass) {
+                Path classFile = getClassFile(javaClass);
+                if (classFile == null) {
+                    return false;
+                }
+                return allPlatformDirs.stream().anyMatch(classFile::startsWith);
```

Comment:
💅 It looks like at least some of these stream searches may benefit from being extracted in a named function. This can be something like `isInSomeDirFrom(allPlatformDirs, classFile)`.

Reply:
👍 done for this one and others